astar_path_planner/src/astar_path_planner.cpp

```diff
@@ -0,0 +1,186 @@
+  std::vector<Point> initial_path = {start};
+  frontier.push({initial_path, GetPathCost(initial_path)});
+
+  while (!frontier.empty()) {
+    const auto [path, cost] = frontier.top();
+    frontier.pop();
+    const auto last_state = path.back();
+
+    if (expanded.find(last_state) == expanded.end()) {
+      expanded.insert(last_state);
+
+      if (IsGoal(last_state)) {
+        std::vector<Point> reduced_path{path};
+        ReducePath(reduced_path);
+        return reduced_path;
+      }
+
+      const auto neighbors = GetAdjacentPoints(last_state);
+
+      for (const auto & neighbor : neighbors) {
+        std::vector<Point> new_path(path);
+        new_path.push_back(neighbor);
+        const auto cost = GetPathCost(new_path);
+        frontier.push({new_path, cost});
+      }
+    }
+  }
+
+  throw std::runtime_error("Could not find path!");
+}
+
+bool AStarPathPlanner::IsGoal(const Point & point)
+{
+  return (point - goal_).norm() < goal_threshold_;
+}
+
+std::vector<Point> AStarPathPlanner::GetAdjacentPoints(const Point & point)
+{
+  std::vector<Point> neighbors;
+
+  for (auto dx = -grid_size_; dx <= grid_size_; dx += grid_size_) {
+    for (auto dy = -grid_size_; dy <= grid_size_; dy += grid_size_) {
+      if (std::abs(dx) < 1e-4 && std::abs(dy) < 1e-4) {
+        continue;
+      }
+      const Point neighbor{point.x() + dx, point.y() + dy};
+      if (!PointInCollision(neighbor)) {
+        neighbors.push_back(neighbor);
+      }
+    }
+  }
+  return neighbors;
+}
+
+double AStarPathPlanner::GetPathCost(const std::vector<Point> & path)
+{
+  // Distance between points
+  std::vector<Eigen::Vector2d> edges;
+  std::adjacent_difference(path.begin(), path.end(), std::back_inserter(edges));
+  double cost = std::accumulate(
+    edges.begin(), edges.end(), 0.0, [](const double & total, const Eigen::Vector2d & edge) {
+      return total + edge.norm();
+    });
+
+  // Heuristic
+  cost += (path.back() - goal_).norm();
+
+  return cost;
+}
+
+bool AStarPathPlanner::PointInCollision(const Point & point)
+{
+  const auto footprint = ros_costmap_->getRobotFootprintPolygon();
+
+  std::vector<nav2_costmap_2d::MapLocation> footprint_in_map;
+
+  auto transform_footprint_point = [&point](
+    const auto & footprint_point) {
+      nav2_costmap_2d::MapLocation location;
+      location.x = footprint_point.x + point.x();
+      location.y = footprint_point.y + point.y();
+      return location;
+    };
+
+  std::transform(
+    footprint.points.begin(), footprint.points.end(), std::back_inserter(
+      footprint_in_map), transform_footprint_point);
+
+  auto costmap = ros_costmap_->getCostmap();
+
+  std::vector<nav2_costmap_2d::MapLocation> map_cells;
+  costmap->convexFillCells(footprint_in_map, map_cells);
+
+  return std::any_of(
+    map_cells.begin(), map_cells.end(), [costmap](const auto & cell) {
+      const auto cost = costmap->getCost(cell.x, cell.y);
+      return cost == 253;
+    });
+}
+
+template<typename InputIterator>
```

Comment:
Why is this templated? It is always a Point iterator right?

Reply:
Yeah, it'll always be a `std::vector<Point>::iterator` in this case. This is mostly written this way out of habit, because this is how you're supposed to write container-agnostic algorithms.